File: 12-2.js
const example = `Sabqponm
abcryxxl
accszExk
acctuvwj
abdefghi`;

const all =
    (...fs) =>
    x =>
        fs.every(f => f(x));

const compose = (...fs) => {
    fs.reverse();
    return x => {
        for (const f of fs) x = f(x);
        return x;
    };
};

const collect = (acc, elem) => [...acc, elem];
const replace = (acc, elem) => elem;

const a = "a".charCodeAt(0);
const diff = c => c.charCodeAt(0) - a;
const norm = c => (c === "S" ? "a" : c === "E" ? "z" : c);

const isStart = c => c === "a";
const isEnd = c => c === "E";

const height = compose(diff, norm);
const heights = row => row.map(height);
const initMap = grid => grid.map(heights);

const infinity = _ => Infinity;
const initSteps = row => row.map(infinity);
const initVisited = map => map.map(initSteps);

const sameRow = target => pos => pos.row === target.row;
const sameCol = target => pos => pos.col === target.col;
const arrived = target => all(sameRow(target), sameCol(target));

const visit = visited => (pos, steps) => (visited[pos.row][pos.col] = steps);

const shorter = visited => steps => pos =>
    visited[pos.row][pos.col] > steps + 1;

const inbound = map => pos =>
    0 <= pos.row &&
    pos.row < map.length &&
    0 <= pos.col &&
    pos.col < map[0].length;

const reachable = map => from => to =>
    map[to.row][to.col] - map[from.row][from.col] <= 1;

const visitable = map => from => all(inbound(map), reachable(map)(from));

const neighbors = ({ row, col }) => [
    { row, col: col - 1 },
    { row, col: col + 1 },
    { row: row - 1, col },
    { row: row + 1, col },
];

const reducePos = test => combine => row => (acc, c, col) =>
    test(c) ? combine(acc, { row, col }) : acc;

const collectStarts = reducePos(isStart)(collect);
const collectEnd = reducePos(isEnd)(replace);
const reduceEnd = (end, cs, row) => cs.reduce(collectEnd(row), end);
const reduceStarts = (starts, cs, row) => cs.reduce(collectStarts(row), starts);
const findEnd = grid => grid.reduce(reduceEnd, undefined);
const findStarts = grid => grid.reduce(reduceStarts, []);

const makeWalk = (arrived, visit, visitable, shorter) =>
    function next(pos, steps) {
        visit(pos, steps);

        if (arrived(pos)) return;

        const toVisit = all(visitable(pos), shorter(steps));
        const goVisit = neighbor => next(neighbor, steps + 1);

        neighbors(pos).filter(toVisit).forEach(goVisit);
    };

const lines = data => data.split("\n").slice(0, -1);
const toArray = cs => [...cs];
const makeGrid = data => lines(data).map(toArray);

require("fs").readFile("12.txt", "utf-8", (err, data) => {
    const grid = makeGrid(data);
    const end = findEnd(grid);
    const starts = findStarts(grid);
    const map = initMap(grid);
    const visited = initVisited(map);

    const walk = makeWalk(
        arrived(end),
        visit(visited),
        visitable(map),
        shorter(visited),
    );

    starts.forEach(start => walk(start, 0));

    console.log({
        starts: starts.length,
        end,
        steps: visited[end.row][end.col],
    });
});

File: 12-2.c
#include "math.h"
#include "prelude.h"
#include <stdbool.h>

typedef struct {
	size_t m, n;
} MatDim;

typedef struct {
	size_t row, col;
} Pos;

static const char *example = "Sabqponm\n"
			     "abcryxxl\n"
			     "accszExk\n"
			     "acctuvwj\n"
			     "abdefghi\n";

static MatDim matrix_dimensions(const char *input) {
	MatDim md = {0};
	for (char c; (c = *input++);) {
		if (c == '\n') md.m++;
		if (!md.m) md.n++;
	}
	return md;
}

static size_t idx(Pos pos, MatDim md) {
	return pos.row * md.n + pos.col;
}

static void walk(Pos current, Pos target, int steps, MatDim md, char *map,
		 int *visited) {
	size_t ci = idx(current, md);
	visited[ci] = steps;
	if (current.row == target.row && current.col == target.col) return;

	Pos neighbors[4];
	size_t n = 0;
	if (current.row > 0) {
		neighbors[n] = current;
		neighbors[n++].row--;
	}
	if (current.row < md.m - 1) {
		neighbors[n] = current;
		neighbors[n++].row++;
	}
	if (current.col > 0) {
		neighbors[n] = current;
		neighbors[n++].col--;
	}
	if (current.col < md.n - 1) {
		neighbors[n] = current;
		neighbors[n++].col++;
	}

	for (size_t i = 0; i < n; i++) {
		Pos next = neighbors[i];
		size_t ni = idx(next, md);

		bool reachable = map[ni] - map[ci] <= 1;
		if (!reachable) continue;
		bool shorter = visited[ni] != -1 && visited[ni] <= steps + 1;
		if (shorter) continue;

		walk(next, target, steps + 1, md, map, visited);
	}
}

static int solve(char *input) {
	const MatDim md = matrix_dimensions(input);

	char *map = malloc(sizeof(char[md.m * md.n]));
	int *visited = malloc(sizeof(int[md.m * md.n]));
	Pos *starts = malloc(sizeof(Pos[md.m * md.n]));

	Pos end = {0};
	size_t n_starts = 0;
	for (size_t row = 0; row < md.m; row++) {
		for (size_t col = 0; col < md.n; col++) {
			const size_t i = row * md.n + col;
			const char c = input[row * (md.n + 1) + col];
			map[i] = c == 'S' ? 'a' : c == 'E' ? 'z' : c;
			visited[i] = -1;
			if (c == 'S' || c == 'a') starts[n_starts++] = (Pos){.row = row, .col = col};
			if (c == 'E') end = (Pos){.row = row, .col = col};
		}
	}

	for (size_t i = 0; i < n_starts; i++)
		walk(starts[i], end, 0, md, map, visited);
	int steps = visited[end.row * md.n + end.col];

	free(starts);
	free(map);
	free(visited);

	return steps;
}

int main(void) {
	RUN(solve, strdup(example));
	RUN(solve, read_to_string("12.txt"));

	return 0;
}
